File: ui/goth/primitives/resizable.go
package primitives

import (
	"strconv"

	"github.com/a-h/templ"
)

// Resizable (P61, family F4) is a split-pane layout with a draggable/keyboard
// separator. The caller composes three parts in order: a primary ResizablePane,
// a ResizableHandle, and a secondary ResizablePane.
//
// No-JS baseline. The server owns the default split: the root carries
// data-default-size (the primary pane's percentage) and external component CSS
// maps it to the primary pane's flex-basis through the --goth-resize-basis custom
// property, so the split renders at the server-chosen ratio with no JavaScript.
// The panes are role=group regions; the handle is a role=separator with
// aria-valuenow/min/max and aria-controls.
//
// Enhancement (gothResizable, §8 controller, ratified by the 2026-07-18
// gate-b-review.md addendum). The controller reads pointer-drag (with pointer
// capture) and APG window-splitter keyboard input (Arrow keys, Home, End; RTL-aware
// per the roving.js precedent), clamps to the min/max bounds, and writes the new
// geometry through the controller-owned CSSOM custom property
// --goth-resize-basis — NEVER a server-rendered style attribute (README §4 frozen
// invariant). It updates the handle's aria-valuenow and dispatches goth:change.
//
// data-slot hooks: resizable, resizable-pane, resizable-handle.

// ResizableOrientation is the typed pane-arrangement enum. The zero value arranges
// the panes horizontally (side by side, a vertical separator); vertical stacks them
// (a horizontal separator).
type ResizableOrientation string

const (
	// ResizableHorizontal is the zero value and the documented default.
	ResizableHorizontal ResizableOrientation = ""
	ResizableVertical   ResizableOrientation = "vertical"
)

// Valid reports whether o is a known ResizableOrientation.
func (o ResizableOrientation) Valid() bool {
	switch o {
	case ResizableHorizontal, ResizableVertical:
		return true
	default:
		return false
	}
}

func (o ResizableOrientation) attr() string {
	if o == ResizableVertical {
		return "vertical"
	}
	return "horizontal"
}

// separatorOrientation is the aria-orientation of the handle: a horizontal pane
// arrangement uses a vertical separator, and vice versa.
func (o ResizableOrientation) separatorOrientation() string {
	if o == ResizableVertical {
		return "horizontal"
	}
	return "vertical"
}

// clampSize clamps v into [min, max].
func clampSize(v, min, max int) int {
	if v < min {
		return min
	}
	if v > max {
		return max
	}
	return v
}

// snap5 snaps v to the nearest multiple of 5 within [5, 95]. The server default
// split is expressed through discrete data-default-size buckets so external CSS
// (not a server-rendered inline style) can render the initial geometry.
func snap5(v int) int {
	v = ((v + 2) / 5) * 5
	return clampSize(v, 5, 95)
}

// ResizableProps configures the Resizable root (P61). The zero value is a valid
// horizontal 50/50 split with 10/90 bounds.
type ResizableProps struct {
	Base
	// Orientation selects the pane arrangement. Zero value is horizontal.
	Orientation ResizableOrientation
	// DefaultSize is the primary pane's percentage of the container for the no-JS
	// baseline (snapped to the nearest 5, clamped to [5,95]). Zero value is 50.
	// Keep it equal to the ResizableHandle's Value so the AT value and the rendered
	// geometry agree.
	DefaultSize int
}

func (p ResizableProps) defaultSize() int {
	v := p.DefaultSize
	if v <= 0 {
		v = 50
	}
	return snap5(v)
}

// ResizablePaneProps configures one ResizablePane (a scrollable content region).
type ResizablePaneProps struct {
	Base
	// Label is the pane's accessible name (aria-label). Recommended so a separator's
	// aria-controls target is nameable and the region is announced.
	Label string
}

// ResizableHandleProps configures the ResizableHandle (the role=separator between
// the panes).
type ResizableHandleProps struct {
	Base
	// Orientation must match the root's Orientation so the separator's
	// aria-orientation is correct. Zero value is horizontal (a vertical separator).
	Orientation ResizableOrientation
	// Value is the primary pane's current percentage (aria-valuenow). Zero value is
	// 50. Keep it equal to the root's DefaultSize.
	Value int
	// Min and Max are the resize bounds (aria-valuemin/valuemax). Zero values are
	// 10 and 90.
	Min int
	Max int
	// ControlsID is the id of the primary pane the separator resizes (aria-controls).
	// Recommended so assistive technology links the separator to the pane it sizes.
	ControlsID string
	// Label is the separator's accessible name (aria-label). Empty renders the
	// default "Resize".
	Label string
}

func (p ResizableHandleProps) bounds() (value, min, max int) {
	min = p.Min
	if min <= 0 {
		min = 10
	}
	max = p.Max
	if max <= 0 || max > 100 {
		max = 90
	}
	if max < min {
		min, max = max, min
	}
	value = p.Value
	if value <= 0 {
		value = 50
	}
	return clampSize(value, min, max), min, max
}

func resizableClass(p ResizableProps) string { return classNames("goth-resizable", p.Class) }

func resizableAttrs(p ResizableProps) templ.Attributes {
	owned := templ.Attributes{
		"data-slot":         "resizable",
		"data-orientation":  p.Orientation.attr(),
		"data-default-size": strconv.Itoa(p.defaultSize()),
		"x-data":            "gothResizable",
	}
	return ownedAttrs(p.Base, owned)
}

func resizablePaneClass(p ResizablePaneProps) string {
	return classNames("goth-resizable-pane", p.Class)
}

func resizablePaneAttrs(p ResizablePaneProps) templ.Attributes {
	owned := templ.Attributes{
		"data-slot": "resizable-pane",
		"role":      "group",
	}
	if p.Label != "" {
		owned["aria-label"] = p.Label
	}
	return ownedAttrs(p.Base, owned)
}

func resizableHandleClass(p ResizableHandleProps) string {
	return classNames("goth-resizable-handle", p.Class)
}

func resizableHandleAttrs(p ResizableHandleProps) templ.Attributes {
	value, min, max := p.bounds()
	label := p.Label
	if label == "" {
		label = "Resize"
	}
	owned := templ.Attributes{
		"data-slot":        "resizable-handle",
		"role":             "separator",
		"tabindex":         "0",
		"aria-orientation": p.Orientation.separatorOrientation(),
		"aria-valuenow":    strconv.Itoa(value),
		"aria-valuemin":    strconv.Itoa(min),
		"aria-valuemax":    strconv.Itoa(max),
		"aria-label":       label,
	}
	if p.ControlsID != "" {
		owned["aria-controls"] = p.ControlsID
	}
	return ownedAttrs(p.Base, owned)
}

File: ui/goth/assets/src/js/controllers/resizable.js
// gothResizable controller (GOTH-6.3; §8 eleventh controller, ratified by the
// 2026-07-18 gate-b-review.md addendum).
//
// Backs Resizable (P61). The server owns the default split (the root's
// data-default-size + external CSS render the initial geometry through the
// --goth-resize-basis custom property with NO JavaScript). This controller adds the
// live resize: pointer drag (with pointer capture) and APG window-splitter keyboard
// input (Arrow keys, Home, End), clamped to the handle's aria-valuemin/valuemax
// bounds. It writes the new geometry through the controller-owned CSSOM custom
// property `--goth-resize-basis` on the root — never a server-rendered style
// attribute (README §4 frozen invariant) — and keeps the handle's aria-valuenow in
// sync. Horizontal arrow direction is RTL-aware per the roving.js precedent.
//
// It holds no domain store: the numeric split percentage is local interaction state
// only; the server owns the authoritative default.

// keyStep: percentage moved per Arrow keypress (APG small increment).
const keyStep = 5;

function isRTL(el) {
  return (
    typeof getComputedStyle === "function" &&
    getComputedStyle(el).direction === "rtl"
  );
}

function toInt(v, fallback) {
  const n = parseInt(v, 10);
  return Number.isFinite(n) ? n : fallback;
}

export default function gothResizable() {
  return {
    _root: null,
    _handle: null,
    _vertical: false,
    _basis: 50,
    _min: 10,
    _max: 90,
    _dragging: false,
    _startPos: 0,
    _startBasis: 50,
    _pointerId: null,

    init() {
      // Cache the root at init; a method invoked from a descendant's handler sees
      // $el bound to that descendant (the GOTH-4.1 descendant-$el discipline).
      this._root = this.$el;
      this._handle = this._root.querySelector('[data-slot="resizable-handle"]');
      if (!this._handle) return;

      this._vertical = this._root.getAttribute("data-orientation") === "vertical";
      this._min = toInt(this._handle.getAttribute("aria-valuemin"), 10);
      this._max = toInt(this._handle.getAttribute("aria-valuemax"), 90);
      this._basis = this._clamp(
        toInt(this._handle.getAttribute("aria-valuenow"), 50),
      );

      // Once enhanced the controller owns the geometry: sync the custom property to
      // aria-valuenow so JS and the no-JS bucket CSS agree even at odd granularities.
      this._root.setAttribute("data-enhanced", "true");
      this._apply();

      this._onPointerDown = (e) => this._pointerDown(e);
      this._onPointerMove = (e) => this._pointerMove(e);
      this._onPointerUp = (e) => this._pointerUp(e);
      this._onKeyDown = (e) => this._keyDown(e);

      this._handle.addEventListener("pointerdown", this._onPointerDown);
      this._handle.addEventListener("pointermove", this._onPointerMove);
      this._handle.addEventListener("pointerup", this._onPointerUp);
      this._handle.addEventListener("keydown", this._onKeyDown);
    },

    _clamp(v) {
      return Math.max(this._min, Math.min(this._max, v));
    },

    // _apply writes the current split through the CSSOM custom property (never a
    // server-rendered style attribute) and mirrors aria-valuenow for AT.
    _apply() {
      this._root.style.setProperty("--goth-resize-basis", this._basis + "%");
      this._handle.setAttribute("aria-valuenow", String(Math.round(this._basis)));
    },

    _setBasis(v) {
      const next = this._clamp(v);
      if (next === this._basis) return;
      this._basis = next;
      this._apply();
      this.$dispatch("goth:change", { size: Math.round(this._basis) });
    },

    _containerSize() {
      return this._vertical ? this._root.clientHeight : this._root.clientWidth;
    },

    _pointerDown(e) {
      this._dragging = true;
      this._pointerId = e.pointerId;
      if (this._handle.setPointerCapture) {
        this._handle.setPointerCapture(e.pointerId);
      }
      this._startPos = this._vertical ? e.clientY : e.clientX;
      this._startBasis = this._basis;
      e.preventDefault();
    },

    _pointerMove(e) {
      if (!this._dragging) return;
      const pos = this._vertical ? e.clientY : e.clientX;
      let deltaPx = pos - this._startPos;
      // RTL flips the horizontal axis: the primary pane sits on the inline-end
      // (right) side, so a rightward drag shrinks it.
      if (!this._vertical && isRTL(this._root)) deltaPx = -deltaPx;
      const size = this._containerSize();
      if (size <= 0) return;
      this._setBasis(this._startBasis + (deltaPx / size) * 100);
    },

    _pointerUp(e) {
      if (!this._dragging) return;
      this._dragging = false;
      if (this._handle.releasePointerCapture && this._pointerId !== null) {
        try {
          this._handle.releasePointerCapture(this._pointerId);
        } catch (_) {
          /* capture already released */
        }
      }
      this._pointerId = null;
    },

    _keyDown(e) {
      const rtl = !this._vertical && isRTL(this._root);
      let handled = true;
      switch (e.key) {
        case "ArrowRight":
          // Move the separator visually right; in RTL the primary pane is on the
          // right, so the same visual move shrinks it.
          if (this._vertical) handled = false;
          else this._setBasis(this._basis + (rtl ? -keyStep : keyStep));
          break;
        case "ArrowLeft":
          if (this._vertical) handled = false;
          else this._setBasis(this._basis + (rtl ? keyStep : -keyStep));
          break;
        case "ArrowDown":
          if (this._vertical) this._setBasis(this._basis + keyStep);
          else handled = false;
          break;
        case "ArrowUp":
          if (this._vertical) this._setBasis(this._basis - keyStep);
          else handled = false;
          break;
        case "Home":
          this._setBasis(this._min);
          break;
        case "End":
          this._setBasis(this._max);
          break;
        default:
          handled = false;
      }
      if (handled) e.preventDefault();
    },
  };
}
